### PR TITLE
Added criteria API logic

### PR DESCRIPTION
**Pull request by artemklishch**

### CODE REVIEW

File: src/main/java/org/example/intro/controller/BookController.java [hunk 1]
```diff
@@ -34,4 +35,9 @@ public BookDto findById(@PathVariable Long id) {
     public void delete(@PathVariable Long id) {
         bookService.delete(id);
     }
+
+    @GetMapping("/search")
+    public List<BookDto> search(BookSearchParametersDto searchParameters) {
```
Comment (Elena-Bruyako) [inline src/main/java/org/example/intro/controller/BookController.java]: Add Pageable

File: src/main/java/org/example/intro/repository/BookSpecificationBuilder.java [hunk 2]
```diff
@@ -0,0 +1,36 @@
...
+public class BookSpecificationBuilder implements SpecificationBuilder<Book> {
+    private final SpecificationProviderManager<Book> bookBookSpecificationProviderManager;
+
+    @Override
+    public Specification<Book> build(BookSearchParametersDto searchParameters) {
+        System.out.println("searchParameters: " + searchParameters);
+        Specification<Book> spec = Specification.where(null);
+        if (searchParameters.titles() != null && searchParameters.titles().length > 0) {
+            spec = spec.and(bookBookSpecificationProviderManager
+                    .getSpecificationProvider("title")
```
Comment (Elena-Bruyako) [inline src/main/java/org/example/intro/repository/BookSpecificationBuilder.java]: Create constants for title, author,  isbn

File: src/main/java/org/example/intro/repository/book/spec/AuthorSpecificationProvider.java [hunk 3]
```diff
@@ -0,0 +1,21 @@
...
+import org.example.intro.model.Book;
+import org.example.intro.repository.SpecificationProvider;
+import org.springframework.data.jpa.domain.Specification;
+import org.springframework.stereotype.Component;
+
+@Component
+public class AuthorSpecificationProvider implements SpecificationProvider<Book> {
+    @Override
+    public String getKey() {
+        return "author";
```
Comment (Elena-Bruyako) [inline src/main/java/org/example/intro/repository/book/spec/AuthorSpecificationProvider.java]: Create constant

File: src/main/java/org/example/intro/repository/book/spec/ISBNSpecificationProvider.java [hunk 4]
```diff
@@ -0,0 +1,21 @@
...
+import org.example.intro.model.Book;
+import org.example.intro.repository.SpecificationProvider;
+import org.springframework.data.jpa.domain.Specification;
+import org.springframework.stereotype.Component;
+
+@Component
+public class ISBNSpecificationProvider implements SpecificationProvider<Book> {
+    @Override
+    public String getKey() {
+        return "isbn";
```
Comment (Elena-Bruyako) [inline src/main/java/org/example/intro/repository/book/spec/ISBNSpecificationProvider.java]: Create constant

File: src/main/java/org/example/intro/repository/book/spec/TitleSpecificationProvider.java [hunk 5]
```diff
@@ -0,0 +1,21 @@
...
+import org.example.intro.model.Book;
+import org.example.intro.repository.SpecificationProvider;
+import org.springframework.data.jpa.domain.Specification;
+import org.springframework.stereotype.Component;
+
+@Component
+public class TitleSpecificationProvider implements SpecificationProvider<Book> {
+    @Override
+    public String getKey() {
+        return "title";
```
Comment (Elena-Bruyako) [inline src/main/java/org/example/intro/repository/book/spec/TitleSpecificationProvider.java]: Create constant

**Review by Elena-Bruyako — CHANGES_REQUESTED** [review]
Good job! Let's improve your solution

File: src/main/java/org/example/intro/repository/book/spec/AuthorSpecificationProvider.java (hunk 3 above)
Comment (artemklishch) [reply src/main/java/org/example/intro/repository/book/spec/AuthorSpecificationProvider.java]: ++

File: src/main/java/org/example/intro/repository/BookSpecificationBuilder.java (hunk 2 above)
Comment (artemklishch) [reply src/main/java/org/example/intro/repository/BookSpecificationBuilder.java]: ++

File: src/main/java/org/example/intro/repository/book/spec/ISBNSpecificationProvider.java (hunk 4 above)
Comment (artemklishch) [reply src/main/java/org/example/intro/repository/book/spec/ISBNSpecificationProvider.java]: ++

File: src/main/java/org/example/intro/repository/book/spec/TitleSpecificationProvider.java (hunk 5 above)
Comment (artemklishch) [reply src/main/java/org/example/intro/repository/book/spec/TitleSpecificationProvider.java]: ++

File: src/main/java/org/example/intro/controller/BookController.java (hunk 1 above)
Comment (artemklishch) [reply src/main/java/org/example/intro/controller/BookController.java]: ++

**Pull request merged**